akka-http-core/src/main/scala/akka/http/scaladsl/DiscoveryClientTransport.scala

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.http.scaladsl
+
+import java.net.InetSocketAddress
+
+import akka.actor.ActorSystem
+import akka.annotation.ApiMayChange
+import akka.discovery.ServiceDiscovery
+import akka.dispatch.ExecutionContexts
+import akka.http.scaladsl.Http.OutgoingConnection
+import akka.http.scaladsl.settings.ClientConnectionSettings
+import akka.stream.{ AbruptStageTerminationException, Attributes, FlowShape, Inlet, Outlet }
+import akka.stream.scaladsl.{ Flow, Keep, Source, Tcp }
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler }
+import akka.util.{ ByteString, OptionVal }
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.{ Failure, Success, Try }
+
+@ApiMayChange
+class DiscoveryClientTransport(discovery: ServiceDiscovery) extends ClientTransport {
+  override def connectTo(host: String, port: Int, settings: ClientConnectionSettings)(implicit system: ActorSystem): Flow[ByteString, ByteString, Future[Http.OutgoingConnection]] = {
+    // The current implementation invokes service discovery for every connection. That might be heavy for
+    // some discovery mechanisms. We might want to eventually move service discovery out of the 'hot path'
+    // with a more push-based API for refreshing discovery information, see
+    // https://github.com/akka/akka/issues/28931
+    implicit val ec = system.dispatcher
+    futureFlow(
+      discovery.lookup(host, settings.connectingTimeout).map { resolved =>
+        resolved.addresses match {
+          case Seq() =>
+            throw new IllegalStateException(s"No addresses looking up [$host]")
+          case addresses =>
+            // TODO take an arbitrary address? Or use some strategy?
+            val address = addresses(0)
+            address.address match {
+              case Some(inetAddress) =>
+                Tcp().outgoingConnection(new InetSocketAddress(inetAddress, address.port.getOrElse(port)), settings.localAddress,
+                  settings.socketOptions, halfClose = true, settings.connectingTimeout, settings.idleTimeout)
+                  .mapMaterializedValue(_.map(tcpConn => OutgoingConnection(tcpConn.localAddress, tcpConn.remoteAddress))(system.dispatcher))
+              case None =>
+                Tcp().outgoingConnection(InetSocketAddress.createUnresolved(address.host, address.port.getOrElse(port)), settings.localAddress,
+                  settings.socketOptions, halfClose = true, settings.connectingTimeout, settings.idleTimeout)
+                  .mapMaterializedValue(_.map(tcpConn => OutgoingConnection(tcpConn.localAddress, tcpConn.remoteAddress))(system.dispatcher))
+            }
+        }
+      }
+    ).mapMaterializedValue(_.flatten)
+
+  }
+
+  private def futureFlow[I, O, M](flow: Future[Flow[I, O, M]]): Flow[I, O, Future[M]] =
```

Comment:
Could we use the 2.5 `Flow.lazyInitAsync` instead?

Reply:
I tried that initially. That materializes a future that is resolved when the first element arrives. By flatmapping that future I can keep the `ClientTransport` signature intact, but then a `singleRequest` call stalls. I haven't dug deeper, but perhaps we don't send any elements before the materialized future completes?

Reply:
`NewHostConnectionPool` uses this future to trigger `onConnectionAttemptSucceeded`, after which the slot becomes `Idle` for accepting requests.

Reply:
Makes sense, forgot that `lazyInitAsync` waits for the first element.